cranelift-codegen/meta/src/isa/x86/legalize.rs

```diff
@@ -290,4 +295,61 @@ pub fn define(shared: &mut SharedDefinitions, x86_instructions: &InstructionGrou
     );
 
     group.build_and_add_to(&mut shared.transform_groups);
+
+    let mut narrow = TransformGroupBuilder::new(
+        "x86_narrow",
+        r#"
+    Legalize instructions by narrowing.
+
+    Use x86-specific instructions if needed."#,
+    )
+    .isa("x86")
+    .chain_with(shared.transform_groups.by_name("narrow").id);
+
+    // SIMD
+    let broadcast_immediate_u8 = Literal::constant(uimm8, 0x00);
+    let broadcast_immediate_f64 = Literal::constant(ieee64, 0x00);
+    let lane_zero = Literal::constant(uimm8, 0x00);
+    let lane_one = Literal::constant(uimm8, 0x01);
+
+//    narrow.legalize(
+//        def!(y = splat.I8x16(x)), // TODO this is not possible yet
+//        vec![
+//            def!(a = bitcast(x)), // move into the lowest 8 bits of an XMM register TODO this encoding does not yet exist
+//            def!(b = f64const(broadcast_immediate_f64)), // setup the shuffle mask in another XMM register; this only works if f64const zeroes out the top 64 bits TODO verify this mask is correct
+//            def!(y = x86_pshufb(a, b)), // PSHUFB takes two XMM operands, one of which is a shuffle mask (i.e. b); TODO this encoding does not exist yet
+//        ],
+//    );
+
+//    narrow.legalize(
+//        def!(y = splat.I16x8(x)), // TODO this is not possible yet
+//        vec![
+//            def!(a = bitcast(x)), // move into the lowest 16 bits of an XMM register TODO this encoding does not yet exist
+//            def!(b = x86_pshuf(a, broadcast_immediate)) // shuffle using PSHUFW (only touches lowest 64 bits, MM width)
+//            def!(y = x86_pshuf(b, different_broadcast_immediate)), // copy lowest 64 bits to highest 64 bits
+//            // TODO this will surely not work because the first pshuf is really using 16x4 types, the second pshuf is using 64x2 types, and then we need y to be 16x8 by the end
+//            // sure would be nice to use VBROADCAST... it's an AVX+ instruction, though
+//        ],
+//    );
+
+    // this works for 32-bit instructions
+    narrow.legalize(
+        def!(y = splat(x)),
+        vec![
+            def!(a = bitcast(x)), // translate to an x86 MOV to get the value in an XMM register
+            def!(y = x86_pshuf(a, broadcast_immediate_u8)) // broadcast the bytes in the XMM register with PSHUF
+        ],
+    );
+
+//    narrow.legalize(
+//        def!(y = splat.I64x2(x)), // TODO this is not possible yet
+//        vec![
+//            def!(a = bitcast(x)), // move into the lowest 64 bits of an XMM register
+//            def!(y = insertlane(a, lane_one, x)), // move into the highest 64 bits of the same XMM register
+//        ],
+//    );
```

Comment:
The majority of my questions are about this unfinished block of code:
 - I started using `bitcast` as a way to move data into the XMM registers; this may be confusing to someone unfamiliar with cranelift. Should this be done differently?
 - the `def!` and `def_rhs!` macros in `cranelift-codegen/meta/src/cdsl/ast.rs` will not allow me to specify vector types--should those macros be changed or is there a different way to do this?
 - is there a way to do the legalizations across all N-bit types (e.g. F16, I16, B16) to avoid repeated code?

Reply:
The key thing with `bitcast` is that the source and destination have to have the same size. The `x86_pshuf` instruction defined elsewhere in this patch states that its input is a vector, so as written, this code is using a `bitcast` to convert between a scalar and a vector with a different size. We should probably add a new instruction named something like `scalar_to_vector` which converts a scalar value into a vector value, filling the other lanes with zeros, as `movq`/`movd` do.

Reply:
For legalizations across F16/I16/B16, we can take advantage of the fact that this meta code is just a Rust program, so we can create a utility routine that takes a `LaneType` argument and defines the legalization with that type.

Reply:
For the `def!` and `def_rhs!` macros, can you say more about what problem you're hitting?

Reply:
- re: `bitcast`: I think what you're saying makes sense; I checked and I think `mov` to an XMM register does zero out the top bits--I'll make the new instruction and call it `scalar_to_vector`. Should it be x86-only?
- re: `def!`: I am trying to tie specific legalizations to specific types; e.g. a splat to an 8x16 is a different sequence of instructions than 16x8 than 32x4, etc. This is rather non-uniform (and I am open to other ways of doing this) but I don't see another way right now because of things like:
    - `pshuf` for 8-bit lanes requiring the shuffle mask in a register (instead of as an immediate),
    - `pshuf` for 64-bit is not existent, presumably because two `pinsr` instructions do the same thing
    - `pshuf` for 16-bit only works on MM registers
But the first problem is that I can't even bind vector types to instructions using `def!`; I don't think Rust's macros can receive something like `splat.I32x4` and know how to split at the `x` (but I may be wrong about this--still new to Rust macros).

Reply:
An update on the last statement "I don't think Rust's macros can receive something like `splat.I32x4` and know how to split at the `x`": I think this may be possible with [procedural macros](https://doc.rust-lang.org/reference/procedural-macros.html) but it would make `def!` a bit more complex and add a `proc_macro` dependency.

Reply:
I think `scalar_to_vector` can be target-independent.

For the `def!` issue, would an option be to add a new `def_vec!` macro which takes additional arguments which provide the vector element and type? That'd be a little redundant, but would allow us to avoid a proc_macro dependency for now. @bnjbvr Do you have any ideas?

Reply:
Re: macro: yes, adding new syntax with special chars could be the way; I too would like to avoid procedural macros, and try to keep the macro as simple as possible. Another way is to bind your instructions for the vector type before using the macro; I've done this when the instruction would need to be bound twice, for instance, so there should be a few examples in the code too.

Reply:
👍 I'm removing the extra macro case and doing the binding prior to using the instruction in `def!`